backend/src/api/routes.ts

```diff
@@ -0,0 +1,32 @@
+import { IncomingMessage, ServerResponse } from "http";
+import {
+    postHandler,
+    getHandler,
+    noResponse,
+} from "../services/requestHandlers";
+import { FriendController, FriendDTO } from "./components/friend/controller";
+import { Friend } from "./components/friend/model/Friend";
+
+export function handler(req: IncomingMessage, res: ServerResponse) {
+    console.log(req.url, req.headers.host)
+    let parsedUrl = new URL(req.url as string, req.headers.host);
+
+    const friendController = new FriendController();
+
+    if (parsedUrl.pathname === "/") {
+        res.writeHead(200, { "Content-Type": "text/plain" });
+        res.end("Hello World");
+    } else if (parsedUrl.pathname === "/friends") {
+        if (req.method === "POST") {
+            postHandler<FriendDTO>(req, res, (f) =>
+                friendController.createNewFriend(f)
+            );
+        } else if (req.method === "GET") {
+            getHandler<Friend>(req, res, parsedUrl, () =>
+                friendController.getFriend(req, parsedUrl)
+            );
+        }
```

Comment:
I think all of this stuff can be done within your FriendController. I'd say managing and handling requests is the controller's main job - remember, controllers are part of the presentation layer, not the business layer. And you are using HTTP to 'present' a UI to your clients. Therefore, your controller (the presentation object) should not be free of HTTP concepts. Your clients use HTTP, so the interface - the controller should use it too. It should speak HTTP and be able to deal with requests - otherwise you just have a business layer object for retrieving and storing entities

Reply:
The only processing of the request that you need to do at this level is to get the path and use that to retrieve the controller with the matching name - then just pass the request on to the controller

Reply:
To put it another way...this 'routes' component's main job is directing requests to the correct handler for the requested web resource (at the moment, only 'friends'). All the handling of the request can be moved to a dedicated component. This component will take the request, then:

1.  Interpret it in terms of the business logic (e.g. "this 'GET friends' request is for a friend object to be retrieved and returned to the caller")
2. Interact with the model or business layer to fulfil that goal
3. Return the result (what was requested) in a way that can be interpret by a HTTP client - a JSON serialised response object, a status code indicating the outcome of the request, any necessary headers etc.

Reply:
thank you - this has been implemented with the HTTP logic moved to the controller. What did you mean by maybe extracting the body of the request - do you mean at the routes level?

Reply:
Cool. So now you've moved that logic into your controller, you can get rid of all references to http concepts in this router. So you don't need to read the http method or hand off to a method-specific handler (e.g. `postHandler`). To fully realise this, you'll need to change your controllers a bit. Basically, you could aim for something that allows to just write:
`if (pathName == 'friends') friendController.handle(request)` . Then you'd just do all interpretation of the friends-specific request within your friends controller handle method. It will look at the request, and if it's a post for example, it will interpret it as, and map it onto, `createFriend`.

Reply:
Then, you can basically get this file into just a switch-case which hands off to the relevant controller based on route

Reply:
please review the routes one more time :)